### ISSUE TITLE
HORRIBLE BUG: The plugin-publish plugin is not documented to be closed-source proprietary software

### ISSUE THREAD

**Issue by Lanchon** [gradle/gradle#13544]
HORRIBLE BUG:

The plugin-publish plugin is not documented to be closed-source proprietary software. The Gradle ecosystem is supposedly free software. The documentation for the plugin-publish plugin is integrated with the rest of the Gradle docs, essentially making users believe that it is trustworthy free software, and leading them to run dangerous and non-free software in their local machines.

The license for this software and its closed-source status should be clearly stated in the docs, so that users can avoid the dangers of non-free software and can comply with entity policies that avoid dependence on non-free software and services.

Now that you already messed up, you REALLY should publish source ASAP. In the meantime, you should clearly warn users about the proprietary nature of this plugin.

**Comment by Vampire** [comment]
Maybe you should try it with a more friendly tone and asking for what you want
instead of using very rude language and demanding what you have no rights for?

The guys at Gradle are doing a great job delivering the best build tool I have ever seen.
Also just that something is closed source does not at all mean it is in any way dangerous.
And at the same time, just that something is open source does not at all mean it is in any way safe.

If they decide to have that plugin being closed-source, that's fine, they have the copyright and can do what they want and they surely have reasons to not open-source that plugin.

You are maybe right, that they could communicate it better.
But then, it is not documented inside the normal Userguide that describes the open-source Gradle.
It is just included in a Guide that shows how to publish to plugin central which you don't have to use if you don't want to.

**Comment by Lanchon** [comment]
ranting is a common means of expression in developer forums. just take a peek into the devel mailing list of the linux kernel for fun. (developers are known for caring more about ideas and less about "feelings".)

rants express frustration and/or anger which exactly express how i feel.

i want to express anger, not a technical problem with a plugin. because i was TRICKED into running non-free software on my systems, a thing that i don't want to do irrespective of all your opinions regarding the "safety" of software. rest assured if the closed-source nature of that software was mentioned ANYWHERE AT ALL, i wouldn't have fallen for that trick.

i'm also angered because i was tricked into LEARNING how to use proprietary tools, which is something i ABSOLUTELY NEVER DO. my investment only happens in free tools: tools that I OWN AND CANNOT BE TAKEN AWAY OR USED TO CONTROL ME.

in essence, i was tricked into INVESTING (my time) in gradle when i didn't want to. and it was not lack of due diligence on my part.

some people may be more careful around gradle if they choose to behave this way. others (like me) tend to sever ties when tricked. that is why i will NEVER be part, for instance, of the .NET ecosystem (which is free as in freedom and technically sound): microsoft breached my trust and they lost me; no amount of promises or turd polishing will have me eating their crap again.

regarding the tone: yes, it's harsh maybe, it's also free speech. the tone is there to alert gradle that this kind of thing is serious for some people.

the whole point is that this issue is trivial and easily prevented. there's nothing trade secret in the plugin, no reason not to publish. conversely, it wouldn't have negatively impacted gradle to document the non-free nature of the plugin portal and its plugin. so why did it happen? because no one in gradle thinks this is an issue. this issue is about changing that. gradle: don't shoot yourselves in the foot for no gain at all!

regarding you: were you offended by the tone? i don't care, being offended is your problem. please keep it out of here if possible, this issue is not about you (or me). you remind me of religious people that get offended by my atheist speech. so what? as richard dawkings says: i'm offended by people who chew gum. so what? and yes, like you, religious people also try to censor me.

but due to your msg i'll state one thing here (which should be obvious because all tracker issues are here to help):
gradle: don't get frustrated by issues like this; get better instead. keep on getting better.

so to recap:
you can argue.
and i can too.
but i was still tricked into using proprietary software against my will.

**Comment by lacasseio** [comment]
Well said @Vampire. I believe this issue should be close as it stands and instead split it into more actionable steps.

- An issue on `gradle/plugin-portal-request` should be open regarding clarifying the language around the `plugin-publish` plugin licenses. Most likely [here](https://plugins.gradle.org/docs/publish-plugin), [here](https://guides.gradle.org/publishing-plugins-to-gradle-plugin-portal/) and [here](https://plugins.gradle.org/plugin/com.gradle.plugin-publish).
- A different issue should be open on `gradle/plugin-portal-request` regarding open-sourcing the `plugin-publish`. There were some talk internally that the plugin is essentially open-source but doesn't have any public repository as a home.
- Another issue should be open on `gradle/plugin-portal-request` for having the option to explicitly declare a plugin as open-source or not and under which license. If my memory serves me right, the plugin portal started off as "assuming" all plugins are open-sourced but nothing enforced this assumption. At this time, there are closed-source and open-source plugin on the portal. It should be clearer for the user to know which one is which. It's also important to mention how it affects the users. A closed source plugin isn't making the users accountable for anything except if you uber-jar the plugin inside your own at which point the licensing is important.
- Finally, another issue should be open on `gradle/plugin-portal-request` to groom the published plugins for any unsafe behaviour to avoid obviously malicious plugins. This feature is a rabbit hole as malicious activity in a plugin can be hard to detect. I believe passing all plugins through a Virus Total scan could be good enough for a "obviously malicious" check.

It's important to keep in mind the [Gradle code of conduct](https://gradle.org/conduct/) for the issue tracking and any other resource of the Gradle community.

**Comment by learnerEvermore** [comment]
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;}  Open source != free.Free != open source.Closed source != !free.Omission != trickery.Documentation issue != product bug.Proprietary = protection of intellectual property.Rant is only as valid as what it claims, otherwise it is just noise. You are using and running (maybe even making) closed source systems all around you. Some are built using Gradle. Some free open-source products are built using Gradle.You may view publicly reviewed code as better in some respects but you have to appreciate that some things put food on people's tables and need to be protected.      Learner Evermore                                                                                                                                                                           From: notifications@github.comSent: June 29, 2020 10:13 a.m.To: gradle@noreply.github.comReply to: reply@reply.github.comCc: subscribed@noreply.github.comSubject: Re: [gradle/gradle] HORRIBLE BUG: The plugin-publish plugin is not documented to be closed-source proprietary software (#13544)  
ranting is a common means of expression in developer forums. just take a peek into the devel mailing list of the linux kernel for fun. (developers are known for caring more about ideas and less about "feelings".)
rants express frustration and/or anger which exactly express how i feel.
i want to express anger, not a technical problem with a plugin. because i was TRICKED into running non-free software on my systems, a thing that i don't want to do irrespective of all your opinions regarding the "safety" of software. rest assured if the closed-source nature of that software was mentioned ANYWHERE AT ALL, i wouldn't have fallen for that trick.
i'm also angered because i was tricked into LEARNING how to use proprietary tools, which is something i ABSOLUTELY NEVER DO. my investment only happens in free tools: tools that I OWN AND CANNOT BE TAKEN AWAY OR USED TO CONTROL ME.
in essence, i was tricked into INVESTING (my time) in gradle when i didn't want to. and it was not lack of due diligence on my part.
some people may be more careful around gradle if they choose to behave this way. others (like me) tend to sever ties when tricked. that is why i will NEVER be part, for instance, of the .NET ecosystem (which is free as in freedom and technically sound): microsoft breached my trust and they lost me; no amount of promises or turd polishing will have me eating their crap again.
regarding the tone: yes, it's harsh maybe, it's also free speech. the tone is there to alert gradle that this kind of thing is serious for some people.
the whole point is that this issue is trivial and easily prevented. there's nothing trade secret in the plugin, no reason not to publish. conversely, it wouldn't have negatively impacted gradle to document the non-free nature of the plugin portal and its plugin. so why did it happen? because no one in gradle thinks this is an issue. this issue is about changing that. gradle: don't shoot yourselves in the foot for no gain at all!
regarding you: were you offended by the tone? i don't care, being offended is your problem. please keep it out of here if possible, this issue is not about you (or me). you remind me of religious people that get offended by my atheist speech. so what? as richard dawkings says: i'm offended by people who chew gum. so what? and yes, like you, religious people also try to censor me.
but due to your msg i'll state one thing here (which should be obvious because all tracker issues are here to help):
gradle: don't get frustrated by issues like this; get better instead. keep on getting better.
so to recap:
you can argue.
and i can too.
but i was still tricked into using proprietary software against my will.

**Comment by ieugen** [comment]
@Lanchon : There are multiple ways of asking for things.
Being demanding, entitled and rude when asking for FREE stuff is not the way to go. 
People have no obligation to do work. for you or anyone else for that matter. 

If you don't like it, express your disapproval and either propose an alternative or use something else that better suits your requirements. 

I don't imagine there is a conspiracy to conquer the world using the proprietary plugin-publish .

**Comment by Lanchon** [comment]
> the plugin portal started off as "assuming" all plugins are open-sourced but nothing enforced this assumption

the current policy is that proprietary closed-source plugins can be published to the portal.

malicious build scripts will happen sooner or later. gradle may want to keep overt closed source plugins out of the portal to increase transparency. companies could still make their secret plugins, they'd just need to host them elsewhere.

repos like f-droid build the stuff they publish themselves and save the source tarball, so you can always access the source of any given version of anything there. of course the source can just pull in a jar with closed source software, but at least that's easy to see. upon detection of such a thing, you'd ban the plugin. the point is that at least those cases can be detected if you look into them.

maybe there's a way to integrate some CI service to do the plugin building for you. just a thought

**Comment by Lanchon** [comment]
> If you don't like it, express your disapproval and either propose an alternative or use something else that better suits your requirements.

have you read the thread? my point is that i was tricked into using this. it is NOT DOCUMENTED not to be free. so are we suppose to ask before using any feature of gradle: hey is this free "too"? are we supposed to clone the source repo and, before using a feature, double check that the source is there? are we supposed to know how the project is split in repos before using it?

gradle dropped the ball on this. and it's a good thing @lacasseio is acknowledging this and proposing to fix it. it gets fixed? good! i shouldn't ask for this to get fixed? nope, disagree.

> People have no obligation to do work. for you or anyone else for that matter.

and you are stating the obvious: everybody knows that people have no obligation of fixing bugs reported in a tracker, including you and me. trackers are still used and work.

jeez! end of being trolled here. no more replies of this kind from me.

**Comment by pioterj** [comment]
I've closed this issue and blocked @Lanchon because of repeated violations of the [code of conduct](https://gradle.org/conduct/) in this issue tracker.

It would be appreciated if someone could create a new issue(s) for this using appropriate language.

**Comment by Vampire** [comment]
This will be my last message to this thread, just to clarify some things you said.

Rants belong to blogs, not to an issue tracker.
If you participate in a community you should follow its code of conduct which you clearly did not do. And even if no code of conduct is defined, you should follow the Netiquette which you also didn't do.

The only thing you gain with such a post as here is to pester the athmosphere of the community.
There are people like Linus that can allow themselves to be gits (his words, not mine), but for most people this is not an acceptable behaviour.

Also you were not tricked into anything.
If you have that strong feelings about closed source software, you need to put more effort in ensuring you only use open source software.
If you for example go to the plugin page on the plugin portal (https://plugins.gradle.org/plugin/com.gradle.plugin-publish) you can see that its id is starting with `com.gradle.`, not `org.gradle.` and `gradle.com` is the domain for the closed source Gradle enterprise package opposed to `gradle.org` which is for the open source part.
Also for each plugin in the portal you find a link to its homepage, in case of the plugin-publish it is https://plugins.gradle.org/docs/publish-plugin and nowhere on that page is any link to the source code.

So yes, the documentation could clearer state that it is not open source, but there is no sign at all that says "this is open source" which I would agree to be tricking you into using it. You just didn't verify enough yourself and request a change in documentation to make it clearer and easier to see. But there is objectively no tricking to complain about.

Freedom of speech is a principle that supports the freedom of an individual or a community to articulate their opinions and ideas without fear of retaliation, censorship, or legal sanction. Freedom of speech does not mean that you can say whatever you like to whomever you like. If you go to your boss and tell him "I don't like you, stupid motherfucker", then you call this "freedom of speech" as long as you like and you can accuse him of censorship if he fires you because of that expression of your opinion, but nevertheless you will get fired and that rightly.

I was not actually offended by the tone, no.
I just think that this is not how people should talk to each other.
And I think that this is not how the people at Gradle deserve to be talked to for providing an amazing tool for free to the world.

And actually, I did not try to censor you.
You can write anything you like in any tone you like, if you can cope with you being ignored or deleted or however Gradle guys will handle the violation of code of conduct.

What I wanted to achieve was helping you being heard and get your issue resolved.
I just advised you on how to improve your communication to actually reach your goal.

If getting the issue fixed actually is your goal that is.
If just pissing people was your goal, then your post was probably appropriate.

Whether you think about what I said or not and whether you heed it in the future
or not is of course totally up to you.

**Comment by pioterj** [comment]
@Vampire @lacasseio @ieugen Thanks for pushing back on this kind of language.
